Closes #1547 bids2 legacy roo tfolder

HenkMutsaerts opened the pull request:
### Linked issue

Closes #1547

HenkMutsaerts committed:
#1547 x.D.ROOT -> x.dir.xASLDerivatives

HenkMutsaerts committed:
#1547 x.D.ROOT -> x.dir.xASLDerivatives nr 2

HenkMutsaerts committed:
#1547 Bugfixes x.D.ROOT -> x.dir.xASLDerivatives

HenkMutsaerts committed:
#1547 xASL_init_DefineStudyData: move parallelization to beginning

HenkMutsaerts committed:
#1547 xASL_init_Iteration: add extra space (cosmetic)

HenkMutsaerts committed:
#1547 Replace others x.D.ROOT -> x.dir.xASLDerivatives

HenkMutsaerts committed:
#1547 xASL_*_BIDS2Legacy: process visits, not subjects

HenkMutsaerts committed:
#1547 xASL_module_BIDS2Legacy: testing multiple runs

HenkMutsaerts committed:
#1547 xASL_imp_DetermineSubjectStructure: separating import & processing

HenkMutsaerts committed:
#1547 ExploreASL_Initialize: further simplification

HenkMutsaerts committed:
#1547 xASL_init_LoadMetadata: Avoid parallel participants.tsv writing

HenkMutsaerts committed:
#1547 remove xASL_imp_UpdateDatasetRoot (separate import & processing)

HenkMutsaerts committed:
#1547 xASL_init_LoadDataPar: harmonization

HenkMutsaerts committed:
#1547 xASL_init_Loaddataset_description: harmonization

HenkMutsaerts committed:
#1547 xASL_init_CreateParticipantsTSV: harmonization

HenkMutsaerts committed:
#1547 Revamp processinig initialization

HenkMutsaerts committed:
#1547 move all variable initializations to xASL_init_SubStructs

HenkMutsaerts committed:
#1547 xASL_init_* fix headers

HenkMutsaerts committed:
#1547 xASL_init_SubjectList: fix subjectRegexp

HenkMutsaerts committed:
#1547 xASL_bids_Add2ParticipantsTSV: minor bugfix n/a

HenkMutsaerts committed:
#1547 xASL_init_Session_TimePoint_Lists: bugfix timepoints

HenkMutsaerts committed:
#1547 Finalize x.ROOT revamp

HenkMutsaerts committed:
#1547 xASL_io_ReadDataPar: bugfix empty structs

HenkMutsaerts committed:
#1547 xASL_module_BIDS2Legacy: simplification

HenkMutsaerts committed:
#1547 Update headers

HenkMutsaerts committed:
#1547 Major bugfix x.SESSIONS

HenkMutsaerts committed:
#1547 xASL_io_CheckDeprecatedFieldsX: only check existing fields

HenkMutsaerts committed:
#1547 major bugfix: loading dataPar.json

HenkMutsaerts committed:
#1547 Bugfix: default to visit _1

HenkMutsaerts committed:
#1547 xASL_qc_CompareTemplate: bugfix too slow pinv

HenkMutsaerts committed:
#1547 xASL_wrp_CreatePopulationTemplates: reduce verbosity for non-existing output

HenkMutsaerts committed:
#1547 xASL_im_CreateGroupAnalysisMask: bugfix loading paths instead of cells

HenkMutsaerts committed:
#1547 xASL_adm_DefineASLSequence: minor bugfix

jan-petr commented on ExploreASL_Initialize.m:
```diff
@@ -342,14 +342,6 @@
-        % We need to check if import/defacing and processing were run separately
-        if x.opts.bImportData && ~x.opts.bProcessData
-            x.opts.bSkipBIDS2Legacy = true;
-            x.opts.bLoadData = false;
```
On Line 817, we still need bLoadData. But that parameter has been removed here. So it crashes on Line 817.

jan-petr replied:
And it's not only there - bLoadData is used on many places before being initialized. It also crashes then in ExploreASL.m line 96.
I do not know why you have removed this from here, but you have to add it then elsewhere.

I talk about bLoadData only, bSkipBIDS2Legacy has been removed everywhere...

jan-petr replied:
Both these errors do not happen when I ran ExploreASL(path, 0, 1) but when running ExploreASL(path, 1, 0).

HenkMutsaerts replied:
bLoadData is a parameter needed for Processing, not Import, which is why I cleaned this. But I see now that it needs proper initialization as false indeed. Fixed in 3572204f7c96a0ba01fba82ca0ac2faac5f3a635

jan-petr commented on Modules/Module_Processing/Module_Population/xASL_adm_GetPopulationSessions.m:
```diff
@@ -1,9 +1,10 @@
 %
 % INPUT:
 %   x                   - struct containing statistical pipeline environment parameters (REQUIRED)
+%   bVerbose            - boolean for verbosity (OPTIONAL, DEFAULT = true)
```
add bVerbose on line 3 as [,bVerbose]

HenkMutsaerts replied:
Fixed in b911e3fbc806998efd455ba2a73018be999120bb

jan-petr reviewed (changes requested):
Minor bugs with large consequences :)

jan-petr commented on Functions/Initialization/xASL_init_Iteration.m:
```diff
@@ -105,17 +94,37 @@
         dbSettings.sets.SUBJECT = SelectedSubjects; % x.SUBJECTS
         dbSettings.x.dir.SUBJECTDIR = fullfile('<ROOT>', '<SUBJECT>');
         dbSettings.x.dir.LockDir = fullfile(dbSettings.x.dir.LockDir, '<SUBJECT>');
+    elseif ~isempty(regexpi(ModName,'BIDS2Legacy', 'once'))
```
Import and BIDS2Legacy options don't differ - should we merge those?

HenkMutsaerts replied:
I would opt to keep them separate, but I agree that this just simplifies reading here. Fixed in 9f57c9fb9a36450260a83fbd59cd9bfb9150cf49

jan-petr commented on Modules/Module_Processing/Module_Population/xASL_im_CreateGroupAnalysisMask.m:
```diff
@@ -72,6 +72,12 @@
         bSkipStandard = true;
+    elseif numel(pathTemplate)>1
+        warning(['Multiple templates found for ' nameTemplate x.S.TemplateNumberName '_bs-mean.nii']);
+        % This situation should not be able to occur
```
Please revise the comment "should not occur" ;)

HenkMutsaerts replied:
Sure. Done in ad47b7f84584d9819f85850317d573152b7144a4

jan-petr commented on Modules/Module_Processing/Module_Population/xASL_im_CreateGroupAnalysisMask.m:
```diff
@@ -72,6 +72,12 @@
+    elseif numel(pathTemplate)>1
+        warning(['Multiple templates found for ' nameTemplate x.S.TemplateNumberName '_bs-mean.nii']);
+        % This situation should not be able to occur
+        eval([nameTemplate ' = ' nameTemplate '{1};']);
```
We shouldn't use eval. Just write nameTemplate = nameTemplate{1}.

1. It's confusing and has no advantage
2. We can't compile statements with `eval`. So no compiled version, no docker.

HenkMutsaerts replied:
1: there is an advantage, nameTemplate is not a variable here containing the template information, but a dummy variable containing template names. Because this is a loop, evalling the same thing 7 times.

2. Tricky thing; I understand.

Now fixed in ad47b7f84584d9819f85850317d573152b7144a4 with a subfunction rather than eval.

jan-petr commented on Modules/Module_Processing/xASL_init_DefineStudyStats.m:
```diff
@@ -0,0 +1,146 @@
+
+% Predefine SETS to avoid empty SETS & import predefined session settings as set settings
+x.S.SetsName{1} = 'session';
+x.S.SetsOptions = '';
```
You initialize as empty char array but later treat this as a cell. Best to just use cell(0), right?

Same with SetsOptions

HenkMutsaerts replied:
Good idea, I didn't know `cell(0)`. I did a quick search and we use this `= '';` initialization for cells quite a lot (probably my mistake). But in fact it doesn't seem to matter, see Matlab screenshot below. So it seems that an empty char array can be indexed by {} cell indices and filled and converted into a cell array. So unless you think initializing a char array is a problem, I would not go looking now to change these throughout the code. Do you agree? I have changed it here and will do when I see it.

    >> filelist = cell(0)
    filelist =
    0×0 empty cell array

    >> filelist{end+1} = 'jan'
    filelist =
    1×1 cell array
    {'jan'}

    >> filelist = ''
    filelist =
    0×0 empty char array

    >> filelist{end+1} = 'jan'
    filelist =
    1×1 cell array
    {'jan'}

Although:

    >> filelist = {''}
    filelist =
    1×1 cell array
    {0×0 char}
    >> filelist{end+1} = 'jan'
    filelist =
    1×2 cell array
    {0×0 char}    {'jan'}

Fixed in 9b26b9648f01ef5bf632e1b2247dda296e6c86e7 (and now also in other functions changed from now in this issue)

jan-petr commented on Modules/xASL_module_BIDS2Legacy.m:
```diff
@@ -77,113 +77,70 @@
 
         %% 2. Define Subject
-        indicesCurrentSubject = find(cellfun( @(y) strcmp(y, x.SUBJECT), {x.modules.bids2legacy.BIDS.subjects.name}));
+        iSubjSess = find(strcmp(x.SUBJECTS, x.SUBJECT));
```
I would strictly try to avoid the use of iSubjSess as a list of sessions. That is - I would always make length of SubjSess equal to nSessions*nSubject without skipping empty sessions. Not sure if you do that or not, but it's not really clear from the code.

That's just a comment, hopefully, the code is correct already.

HenkMutsaerts replied:
You were confused with sessions/runs right? Here, session = a visit. I added a comment to clarify this. OK? Or do we rename some variables?

Fixed in 94563ac2dba3ecaa0f12dfea7912c55e84cc20c3

jan-petr commented on Modules/Module_Processing/xASL_init_Session_TimePoint_Lists.m:
```diff
@@ -0,0 +1,137 @@
+    if isempty(x.SESSIONS)
+        fprintf('%s\n', 'No sessions found, defaulting to a single ASL_1 session');
+        x.SESSIONS{1} = 'ASL_1'; % default session
+	else
```
check identation

HenkMutsaerts replied:
Weird, the indentation looked fine in Matlab. Probably has to do with a tab being four spaces? Actually Matlab indeed complained that it was not aligned, but visually it was aligned. Removing empty spaces and adding tabs fixed it. This is a great example of something that automatic code improvement based on language models could save us time on.

Fixed in 52b2006df0326a4c88d729cb66965978697a61fb (also refactored this function)

jan-petr commented on Modules/Module_Processing/xASL_init_Session_TimePoint_Lists.m:
```diff
@@ -0,0 +1,137 @@
+for iSubj=1:x.dataset.nTotalSubjects
+    iSess=1;
+    iSubjSess = (iSubj-1)*x.dataset.nSessions + iSess;
+    x.dataset.TimePointTotalSubjects{TimePoint(iSubjSess)}{end+1} = x.dataset.TotalSubjects{iSubj};
```
I'm confused if TimePointTotalSubjects is a 1D or 2D cell array. Here it is 2D, below and above 1D.

HenkMutsaerts replied:
I didn't change this behavior, this was just moving code. Now did some major fixes in 52b2006df0326a4c88d729cb66965978697a61fb

jan-petr commented on Modules/Module_Processing/xASL_init_Session_TimePoint_Lists.m:
```diff
@@ -0,0 +1,137 @@
+    x.dataset.TimePointSubjects{iCell} = '';
+end
+    
+for iS=1:x.dataset.nSubjects
```
WHy do we have nSubjects here and nTotalSubjects above? Does it matter?

HenkMutsaerts replied:
Now added this in the comments. nTotalSubjects is before exclusion, nSubjects after.

x.dataset.exclusion is an optional cell list of subjects that you want to exclude from loading (either from rawdata or derivatives)

Fixed in 52b2006df0326a4c88d729cb66965978697a61fb

jan-petr commented on Modules/Module_Processing/xASL_init_Session_TimePoint_Lists.m:
```diff
@@ -0,0 +1,137 @@
+    while ~FoundE % excluded subject not found in previous TimePoint
+        iT=iT+1;  % go to next TimePoint
+        
+        iS  = find(strcmp(x.dataset.TimePointTotalSubjects{iT}, x.dataset.ExcludedSubjects{iE}));
```
what if it doesn't ever find anything - wouldn't this while loop run endlessly?

HenkMutsaerts replied:
Fixed in 47e11499aaa1c09451713efd75adfdfb7106f3be

jan-petr reviewed (changes requested):
A few more minor comments.

HenkMutsaerts committed:
#1547 ExploreASL_Initialize: bLoadData fix

HenkMutsaerts committed:
#1547 xASL_init_Iteration: simplification

HenkMutsaerts committed:
#1547 xASL_adm_GetPopulationSessions: bVerbose in header

jan-petr commented on Modules/Module_Processing/xASL_init_Process.m:
```diff
@@ -20,71 +28,76 @@
+
+            error('Both rawdata and derivatives folders missing, cannot continue processing');
+
+            x.opts.bLoadData = false;
```
This line can't be reached - there's an "error" function called above.

HenkMutsaerts replied:
True. The question remains if x.opts.bLoadData and x.opts.bLoadableData are still usable after my refactoring. Fixed:

`x.opts.DataLoaded`: 509e2976f384163a8f70ac247415a7c4e29b5287
`x.opts.bLoadableData`: a62d07559eac39c6d486820d85acbeaddcf3b67f
`x.opts.bLoadData`: 419395dbfb1c8ae135dce3861dc5d9daa8d9d97e

jan-petr commented on Modules/Module_Processing/xASL_init_Process.m:
```diff
@@ -20,71 +28,76 @@
+    end
     
+    % Verify that the rawdata subfolder exists
+    if ~exist(fullfile(x.dir.DatasetRoot,'rawdata'), 'dir')
```
Above - we have checked that we don't have rawdata folder. But we still try to check this, print a warning and then exit. 
We should skip this check completely when x.opts.ReadRawData is false.

HenkMutsaerts replied:
Yes, now fixed in 419395dbfb1c8ae135dce3861dc5d9daa8d9d97e (or any of my recent other x.opts refactoring)

jan-petr commented on Modules/Module_Processing/xASL_init_SubjectList.m:
```diff
@@ -0,0 +1,181 @@
+    x.dataset.TotalSubjects = x.dataset.ForceInclusionList(:);
+    warning('Using custom list of subjects, on your own risk');
+
+    if ~isfield(x.dataset, 'subjectRegexp')
```
It's really weird that even-though I have dataPar.json in derivatives/ExploreASL directory, and it contains subjectRegexp, it reaches the code here without loading it.

Note that BIDS2Legacy copies dataPar.json from root to derivatives. But processing module is supposed to read derivatives/ExploreASL/dataPar.json

HenkMutsaerts replied:
dataPar.json is a processing settings file, so should be loaded for processing only. BIDS2Legacy doesn't copy dataPar.json anymore, this is now done in 

ExploreASL_Process>xASL_init_DataLoading>xASL_init_LoadDataPar

where it checks for dataPar.json at several locations, in order or preference, and loads it for further processing.

I now changed the order, see issue and new code. Done in latest commits

jan-petr commented on Modules/Module_Processing/xASL_init_Loaddataset_description.m:
```diff
@@ -0,0 +1,102 @@
+if ~isfield(x.dir,'dataset_description')
+	error('Missing dataset_description field...');
+end
+if isempty(x.dir.dataset_description)
```
If you call processing only of derivatives without rawdata, it has empty dataset_description and it crashes here.

HenkMutsaerts replied:
Sorry, this function was simply not finished. Fixed in decce6ef16de421b77588635b54d9b4956c90017

jan-petr commented on Modules/Module_Processing/xASL_init_LoadDataPar.m:
```diff
@@ -0,0 +1,133 @@
+        listDatapar = listRawdata;
+        fprintf('%s\n', ['dataPar.json found in ' fFolder]);
+    else
+        listDatapar = {};
```
In case no dataPar is discovered the it still tries to read it on line 97 and then crashes.

HenkMutsaerts replied:
Oops :)

Fixed in 2105779ffe8d77ebeef8f57150c086e5cbedfb1a

jan-petr reviewed (changes requested):
Few more minor bugs.

HenkMutsaerts committed:
#1547 xASL_im_CreateGroupAnalysisMask: avoid eval

HenkMutsaerts committed:
#1547 xASL_init_Session_TimePoint_Lists: refactor

HenkMutsaerts committed:
#1547 xASL_init_Session_TimePoint_Lists: avoid infinite while-loop

HenkMutsaerts committed:
#1547 xASL_module_BIDS2Legacy: clarify that this does not parse sessions/runs

HenkMutsaerts committed:
#1547 xASL_init_DefineStudyStats: use cell(0)

HenkMutsaerts committed:
#1547 xASL_init_Loaddataset_description: finishing

HenkMutsaerts committed:
#1547 removed redundant x.opts.bLoadData

We already only run these functions if x.opts.bLoadData is true

HenkMutsaerts committed:
#1547 Remove redundant x.opts.bLoadableData

ExploreASL_Initialize initialized this as false, then xASL_init_Process set this to true, if we survive errors for unavailable data folders, then ExploreASL_Process used this. Instead, we now only use the errors in xASL_init_Process

HenkMutsaerts committed:
#1547 xASL_init_*: move x.opts.DataLoaded to correct position

HenkMutsaerts committed:
#1547 Revamp xASL_init_* at beginning ExploreASL_Process

HenkMutsaerts committed:
#1547 xASL_init_* Revamp xASL_init_* at beginning ExploreASL_Process part 2

HenkMutsaerts committed:
#1547 xASL_init_LoadDataPar: bugfix

HenkMutsaerts reviewed (commented):
Some fixes!

jan-petr commented on Modules/Module_Processing/xASL_init_Process.m:
```diff
@@ -10,81 +10,496 @@
 
+%% =======================================================================================================================
+%% =======================================================================================================================
+function [x] = xASL_init_RemoveLockDirs(x)
```
Careful. This can't be a subfunction as it is called also globally!
ExploreASL_Process.m:84:        x = xASL_init_RemoveLockDirs(x);

HenkMutsaerts replied:
Oops I didn't finish this. Fixed in 289f654680dff2b831fd25e67cfb0b9863510072

jan-petr commented on Modules/Module_Processing/xASL_init_DataLoading.m:
```diff
@@ -44,608 +38,56 @@
-    % In a future release we could set all defaults based on a JSON here
-    if ~isfield(x.modules.structural, 'bSegmentSPM12')
-        x.modules.structural.bSegmentSPM12 = false;
+    if x.opts.bReadRawdata && ~xASL_exist(x.dir.RawData, 'dir')
```
If I run Processing module. And derivatives folder exists and rawdata folder is missing. It tries to read rawdata and crashes. Despite that rawdata are not really necessary here.

jan-petr replied:
That's when any processing is executed...

HenkMutsaerts replied:
Sorry for not testing this. Fixed in 136c839b51e258fb1c2b425409b7f0be4898e5ee

jan-petr reviewed (changes requested):
Two small bugs.

HenkMutsaerts committed:
#1547 xASL_init_RemoveLockDirs: move to correct position

HenkMutsaerts committed:
#1547 xASL_init_DataLoading: Finalize x.opts.bReadRawdata management

HenkMutsaerts committed:
#1547 xASL_adm_DefineASLSequence: reduce verbosity

HenkMutsaerts committed:
#1547 xASL_init_Iteration: improve <SESSION> printing

HenkMutsaerts committed:
#1547 xASL_*: reduce verbosity

jan-petr commented on Modules/Module_Processing/xASL_init_DataLoading.m:
```diff
@@ -44,608 +37,46 @@
+            x.opts.bReadRawdata = false;
+            fprintf('%s\n', '/derivatives/ExploreASL folder detected, trying to load subjects from previous ExploreASL processing');
+        else
+            error('No rawdata or derivatives found to load subjects from, first convert dicoms 2 BIDS');
```
Minor error - dicoms->DICOMs. You can immediately resolve after fixing without replying ;)

jan-petr commented on Modules/Module_Processing/xASL_init_DataLoading.m:
```diff
@@ -44,608 +37,46 @@
-        end
     
+    %% 4. Which data to read
+    if ~isfield(x.opts, 'bReadRawdata') || isempty(x.opts.bReadRawdata) || ~islogical(x.opts.bReadRawdata)
```
You probably need to be very careful here. I wouldn't be so strict with the logicals - x.opts.bReadRawdata can still be 0 or 1. Which is then still a boolean in the broad sense, but isn't "logical".

We can probably remove this entire condition as this is an internal variable always being true/false or 1/0 and not a string. Or am I missing something special?

Even for a parameter entered by the user in dataPar.json, I would not be too strict and let them use 1/0 instead of true/false.

HenkMutsaerts replied:
Can be defined in dataPar indeed. But I see your point, will remove the islogical. Fixed in 9afb01be1e81c615a976a2ae525063d7e2e3c0c2

jan-petr commented on Functions/Administration/xASL_adm_DefineASLSequence.m:
```diff
@@ -1,4 +1,4 @@
-function [x] = xASL_adm_DefineASLSequence(x)
+function [x] = xASL_adm_DefineASLSequence(x, bVerbose)
```
Please add bVerbose to line 4, line 10. And if it is an optional parameter, it needs admin on line 23 as well.

HenkMutsaerts replied:
Sorry for missing this. Fixed in 9afb01be1e81c615a976a2ae525063d7e2e3c0c2

jan-petr replied:
Thanks - we still need to list it in the INPUT parameter list...

HenkMutsaerts replied:
Done in 40c014963eebcc2c8d6d01a817980edd83077a5d

jan-petr reviewed (changes requested):
A few small things.

jan-petr commented on Modules/Module_Processing/xASL_init_SubjectList.m:
```diff
@@ -0,0 +1,183 @@
+
+    % And check that this doesn't accidentally contain pipeline directories
+    ListPipelineDir = logical([]);
+    for iSubject=1:x.dataset.nTotalSubjects
```
nTotalSubjects NOT defined at this line! It still has to be calculated from x.dataset.TotalSubjects.

HenkMutsaerts replied:
Fixed in 9afb01be1e81c615a976a2ae525063d7e2e3c0c2

jan-petr reviewed (changes requested):
One more small bug.

HenkMutsaerts committed:
#1547 Small bugfixes

HenkMutsaerts reviewed (commented):
Fixed in 9afb01be1e81c615a976a2ae525063d7e2e3c0c2

HenkMutsaerts committed:
#1547 xASL_adm_DefineASLSequence: final header fix

HenkMutsaerts committed:
#1547 Move /Functions/Import/* -> /Modules/Module_Import/Import/*

HenkMutsaerts committed:
#1547 Move xASL_bids_* to appropriate folders

jan-petr reviewed (approved):
OK
